jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/internal/WebSocketHttpFieldsWrapper.java

```diff
@@ -14,75 +14,118 @@
 package org.eclipse.jetty.websocket.core.server.internal;
 
 import java.util.Collections;
+import java.util.List;
 
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.HttpFields;
 import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpHeaderValue;
 import org.eclipse.jetty.websocket.core.ExtensionConfig;
 import org.eclipse.jetty.websocket.core.server.ServerUpgradeResponse;
 
-public class WebSocketHttpFieldsWrapper extends HttpFieldsWrapper
+public class WebSocketHttpFieldsWrapper extends HttpFields.Mutable.Wrapper
 {
-    private final WebSocketNegotiation _negotiation;
     private final ServerUpgradeResponse _response;
 
     public WebSocketHttpFieldsWrapper(Mutable fields, ServerUpgradeResponse response, WebSocketNegotiation negotiation)
     {
         super(fields);
-        _negotiation = negotiation;
         _response = response;
     }
 
     @Override
-    public boolean onPutField(String name, String value)
+    public HttpField onAddField(HttpField field)
     {
-        if (HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL.is(name))
+        if (field.getHeader() != null)
         {
-            _response.setAcceptedSubProtocol(value);
-            return false;
-        }
+            return switch (field.getHeader())
+            {
+                case SEC_WEBSOCKET_SUBPROTOCOL ->
+                {
+                    _response.setAcceptedSubProtocol(field.getValue());
+                    yield null;
+                }
 
-        if (HttpHeader.SEC_WEBSOCKET_EXTENSIONS.is(name))
-        {
-            _response.setExtensions(ExtensionConfig.parseList(value));
-            return false;
-        }
+                case SEC_WEBSOCKET_EXTENSIONS ->
+                {
+                    _response.addExtensions(ExtensionConfig.parseList(field.getValue()));
+                    yield null;
+                }
 
-        return super.onPutField(name, value);
+                default -> super.onAddField(field);
+            };
+        }
+        return super.onAddField(field);
     }
 
     @Override
-    public boolean onAddField(String name, String value)
+    public boolean onRemoveField(HttpField field)
     {
-        if (HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL.is(name))
+        if (field.getHeader() != null)
         {
-            _response.setAcceptedSubProtocol(value);
-            return false;
+            return switch (field.getHeader())
+            {
+                case SEC_WEBSOCKET_SUBPROTOCOL ->
+                {
+                    _response.setAcceptedSubProtocol(null);
+                    yield false;
+                }
+
+                case SEC_WEBSOCKET_EXTENSIONS ->
+                {
+                    // TODO: why add extensions??
+                    _response.setExtensions(Collections.emptyList());
```

Comment:
This is not strictly correct as it is removing 1 `HttpField` at a time and we could have multiple `SEC_WEBSOCKET_EXTENSIONS` fields.

Maybe we need a `_response.removeExtension(...)` for this.

Reply:
true.    Note that it is mostly not used due to the wrapping of the put methods.  But if an app was to try to do a specific remove....
I have put in a TODO for now, as the problem existed before.   For another PR

Reply:
I don't think this problem did exist before, there was this method which only took the header name, and it was called before removing all `HttpFields` for a `HttpHeader`, not for removing an individual `HttpField`.
```java
public boolean onRemoveField(String name)
```

Reply:
@lachlan-roberts ah the problem before was that the HttpFields wrapper did not handle at all removal of an individual field as it did not modify the iterator.  So it just ignored it.    I'll see if I can fix then....

Reply:
@lachlan-roberts how about this now?

Reply:
Yep that looks like it will work now.